base/setup/usetup/usetup.c

```diff
@@ -626,6 +626,176 @@ UpdateKBLayout(VOID)
     }
 }
 
+/**
+ * @ProgressTimeOutStringHandler
+ * @implemented
+ *
+ * Handles the generation (displaying) of the timeout
+ * countdown to the screen dinamically
+ *
+ * @return
+ * FALSE on function termination
+ *
+ */
+static
+BOOLEAN NTAPI
+ProgressTimeOutStringHandler(
+    IN PPROGRESSBAR Bar,
+    IN BOOLEAN AlwaysUpdate,
+    OUT PSTR Buffer,
+    IN SIZE_T cchBufferSize)
+{
+    ULONG OldProgress = Bar->Progress;
+
+    if (Bar->StepCount == 0)
+    {
+        Bar->Progress = 0;
+    }
+    else
+    {
+        Bar->Progress = Bar->StepCount - Bar->CurrentStep;
+    }
+
+    if (Bar->ProgressFormatText &&
+        (AlwaysUpdate || (Bar->Progress != OldProgress)))
+    {
+        RtlStringCchPrintfA(Buffer, cchBufferSize,
+            Bar->ProgressFormatText, Bar->Progress);
+
+        return TRUE;
+    }
+
+    return FALSE;
+}
+
+/**
+ * @ProgressBarOnReboot
+ * @implemented
+ *
+ * Displays and draws a progress bar with red colour with a countdown.
+ * If the timeout has been reached, it'll go to on the flush page for reboot.
+ *
+ * @return
+ * Nothing
+ *
+ */
+static VOID
+ProgressBarOnReboot(
+    IN PINPUT_RECORD Ir,
+    IN LONG TimeOut)
+{
+    NTSTATUS Status;
+    ULONG StartTime;
+    LONG TimeElapsed;
+    LONG TimerValue = TimeOut, OldTimerValue;
+    LARGE_INTEGER Timeout;
+    PPROGRESSBAR RebootProgressBar;
+    BOOLEAN RefreshProgress = TRUE;
+
+    /* Bail out if the timeout is already zero */
+    if (TimeOut <= 0)
+    {
+        return;
+    }
+
+    /* Create the timeout progress bar and set it up */
+    RebootProgressBar = CreateProgressBarEx(13,
+                                            26,
+                                            xScreen - 13,
+                                            yScreen - 20,
+                                            10,
+                                            24,
+                                            TRUE,
+                                            FOREGROUND_RED | BACKGROUND_BLUE,
+                                            15, 
+                                            NULL,
+                                            MUIGetString(STRING_REBOOTPROGRESSBAR),
+                                            ProgressTimeOutStringHandler);
+
+    StartTime = NtGetTickCount();
+    CONSOLE_Flush();
+
+    OldTimerValue = TimerValue;
+    while (TRUE)
+    {
+        TimeElapsed = NtGetTickCount() - StartTime;
+        if (TimeElapsed >= 1000)
```

Comment:
It would be nice to update counter each 10 ms instead of 1 second, it would make the progress bar movement smoother, almost like in 2nd stage.

Reply:
This behaviour can be expected the same like on 2000/XP USETUPs. Why would want for the progress bar movement be smoother in the first place?

Reply:
Because it would be better. We should not follow Windows designs precisely, when we are not talking about API / interfaces / software compatibility.

Reply:
Why not, but then you should use a "clever" time quantum. No need to use e.g. 1 ms if as the result, the progress-bar square-filling character needs to take much more than 1ms for it to be moved.
A good quantum (using proportionality rules) could be something like, `TimeOut / number_of_squares_making_the_progressbar`.

Reply:
Not sure if I am doing this right -- the bar is accelerating too fast and then it simply stops....
https://pastebin.com/wkDFaE6t

@HBelusca If I understood you right, the IF condition inside the loop is not needed anymore since we want to fill the square bar always and not within a specified interval of time.

Reply:
`StepCount` argument for CreateProgressBarEx is still equal to 15.